include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -3,150 +3,237 @@
 
 #ifdef NDF_PHONG
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_PHONG) {
-#endif
-
-
-			float n = irr_glsl_alpha2_to_phong_exp(a2);
+				CASE_BEGIN(ndf, NDF_PHONG) {
+					float n = irr_glsl_alpha2_to_phong_exp(a2);
 #ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, n, a2);
+					bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG_wo_clamps(NdotH, NdotV2, NdotL2, n, a2);
 #else
-			float ny = irr_glsl_alpha2_to_phong_exp(ay2);
-			bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(currBSDFParams, currInteraction, n, ny, a2, ay2);
-#endif
-
-
-#ifndef ONLY_ONE_NDF
-		} else
+					float ny = irr_glsl_alpha2_to_phong_exp(ay2);
+					bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG_wo_clamps(NdotH, NdotH2, TdotH2, BdotH2, TdotL2, BdotL2, TdotV2, BdotV2, NdotV2, NdotL2, nx, ny, ax2, ay2);
 #endif
+				} CASE_END
 #endif
 
-#ifndef ONLY_ONE_NDF
-		{} //else "empty braces"
-#endif
-	}
+				CASE_OTHERWISE
+				{} //else "empty braces"
+			END_CASES
 
-	uvec3 regs = instr_decodeRegisters(instr);
-#ifdef OP_DIFFUSE
-	if (op==OP_DIFFUSE) {
-		instr_execute_cos_eval_DIFFUSE(instr, regs, params, bsdf_data);
-	} else
-#endif
+				float VdotH = microfacet.isotropic.VdotH;
+				float VdotH_clamp = irr_glsl_conditionalAbsOrMax(is_bsdf, VdotH, 0.0);
+				vec3 fr;
 #ifdef OP_CONDUCTOR
-	if (op==OP_CONDUCTOR) {
-		instr_execute_cos_eval_CONDUCTOR(instr, regs, bxdf_eval_scalar_part, params, bsdf_data);
-	} else
+				if (op == OP_CONDUCTOR)
+					fr = irr_glsl_fresnel_conductor(ior[0], ior[1], VdotH_clamp);
+				else
 #endif
-#ifdef OP_PLASTIC
-	if (op==OP_PLASTIC) {
-		instr_execute_cos_eval_PLASTIC(instr, regs, bxdf_eval_scalar_part, params, bsdf_data);
-	} else
-#endif
-#ifdef OP_COATING
-	if (op==OP_COATING) {
-		instr_execute_cos_eval_COATING(instr, regs, params, bsdf_data);
-	} else
+					fr = irr_glsl_fresnel_dielectric_common(ior2[0], VdotH_clamp);
+
+#ifdef OP_DIELECTRIC
+				if (is_bsdf)
+				{
+					float LdotH = microfacet.isotropic.LdotH;
+					float VdotHLdotH = VdotH * LdotH;
+#ifdef NDF_GGX
+					if (ndf == NDF_GGX)
+						bxdf_eval_scalar_part = irr_glsl_ggx_microfacet_to_light_measure_transform(bxdf_eval_scalar_part, NdotL, refraction, VdotH, LdotH, VdotHLdotH, eta);
+					else
 #endif
-#ifdef OP_DIFFTRANS
-	if (op==OP_DIFFTRANS) {
-		instr_execute_cos_eval_DIFFTRANS(instr, regs, params, bsdf_data);
-	} else
+						bxdf_eval_scalar_part = irr_glsl_microfacet_to_light_measure_transform(bxdf_eval_scalar_part, NdotV, refraction, VdotH, LdotH, VdotHLdotH, eta);
+				}
 #endif
-#ifdef OP_DIELECTRIC
-	if (op==OP_DIELECTRIC) {
-		instr_execute_cos_eval_DIELECTRIC(instr, regs, params, bsdf_data);
-	} else
+				bxdf_eval = fr * bxdf_eval_scalar_part;
+		} else
 #endif
-#ifdef OP_THINDIELECTRIC
-	if (op==OP_THINDIELECTRIC) {
-		instr_execute_cos_eval_THINDIELECTRIC(instr, regs, params, bsdf_data);
-	} else
+		{}
+	}
+
+	bxdf_eval_t result = bxdf_eval;
```

Comment:
can you just do the rest on `bxdf_eval` or just call `bxdf_eval` `result` right away?